### PR TITLE
Mlflow integration

### PR DESCRIPTION
**Pull request by nhorlava**
Key features:

1. Mlflow tracking was added as an additional option for tracking. See `scripts/slurm/mlflow_example/README.md` for an explanation of how to run it. 
**NOTE**: Please re-build the container image before running the code since the container recipe was modified to include mlflow dependency. 

3. Custom Callback for MLflow tracking, as well as mlflow-specific setup function, were implemented and can be found in `./SynthSeg/mlflow_utils.py`. 

	- The Callback inherits from the `mlflow.tensorflow.callback.MLflowCallback`, however, it has key differences: 
		- the WB-like behavior of the `log_freq` argument was implemented: https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/mlflow_utils.py#L12-L17
		- It logs slurm-specific variables (node name, job id, etc. ): https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/mlflow_utils.py#L29
		- One can differentiate between validation and train metrics, as well as between different losses (if the name was specified): https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/mlflow_utils.py#L24 https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/mlflow_utils.py#L51 https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/mlflow_utils.py#L67-L68

4. Mlflow will additionally track the system metrics as specified by the `log_system_metrics=True`: https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/training_with_tfrecords.py#L77
5. To ensure that the experiment run has an appropriate status ("KILLED") if the code was interrupted (e.g. because of hitting the time limit on the cluster), additional signal monitoring was added to the training function: https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/training_with_tfrecords.py#L29-L39 https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/training_with_tfrecords.py#L51-L53

6. To improve reproducibility of neural network training when using mixed precision, the "auto_mixed_precision" argument was set to True (see [here](https://www.tensorflow.org/api_docs/python/tf/config/optimizer/set_experimental_options) ): https://github.com/halirutan/SynthSeg/blob/1fd74e1c8719ad02cd02f60c0739e0ac8ae9a376/SynthSeg/training_with_tfrecords.py#L63

### CODE REVIEW

File: scripts/slurm/mlflow_example/config.ini
Comment (dcfidalgo) [inline scripts/slurm/mlflow_example/config.ini:1]: I would rather make this a shell script (maybe called `mflow_env.sh`), in which you export the necessary [env variables](https://mlflow.org/docs/latest/python_api/mlflow.environment_variables.html) and which you can simply source before the training run (see comment above).

File: SynthSeg/mlflow_utils.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+def set_mlflow_tracking():
+    config = configparser.ConfigParser()
+    config.read(os.environ["MLFLOW_CONFIG"])
+    mlflow.set_tracking_uri(config['HPC_CLOUD']["URI"])
+
+    os.environ['MLFLOW_TRACKING_USERNAME'] = config["LOGIN"]["USERNAME"]
+    os.environ['MLFLOW_TRACKING_PASSWORD'] =  config["LOGIN"]["PASSWORD"]
+    mlflow.set_experiment(config['MLFLOW']["EXPERIMENT"])        
+    mlflow.tensorflow.autolog(disable = True, )
```
Comment (dcfidalgo) [inline SynthSeg/mlflow_utils.py]: I think this function is superfluous since (except for the `autolog` disabling) you can configure everything via [mlflow env variables](https://mlflow.org/docs/latest/python_api/mlflow.environment_variables.html). So instead of defining a `config.ini` + parsing it, I would rather define a shell script and source it before the training (see comment below).

File: SynthSeg/training_options.py [hunk 2]
```diff
@@ -321,6 +321,18 @@ class TrainingOptions(Serializable, OptionsBase):
     If "batch", logs metrics at the end of each batch. If an integer, logs metrics at the end of that
     many batches.
     """
+    
+    mlflow: bool = False
+    """
+    Add a mlflow callback when training with the Dice loss function.
+    """
+
+    mlflow_log_freq: Union[int, str] = "epoch"
```
Comment (dcfidalgo) [inline SynthSeg/training_options.py]: Maybe we could merge this and the `wandb_log_freq` parameter to a general `log_freq` parameter, that both tools use. I don't think there will ever be a use case, where you want to log more/less with one tool than with the other.

File: SynthSeg/training_with_tfrecords.py [hunk 3]
```diff
@@ -16,14 +16,28 @@
 from .metrics_model import WeightedL2Loss, DiceLoss, IdentityLoss, MeanIoU
 from .training_options import TrainingOptions
 from .brain_generator import read_tfrecords
-
+import mlflow
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: I would not make mlflow a hard dependency. Only import it, if `opt.mlflow` is true.

File: SynthSeg/training_with_tfrecords.py [hunk 4]
```diff
@@ -16,14 +16,28 @@
...
 
 class NullStrategy:
     @staticmethod
     def scope():
         return nullcontext()
-
-
+         
+    
+def signal_handler(use_mlflow, signum, frame, ):
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: Since for now we only use signal handling when tracking with mlflow, I would define this handler locally in the `if opts.mlflow` statement, see suggestion below.

File: SynthSeg/training_with_tfrecords.py [hunk 5]
```diff
@@ -16,14 +16,28 @@
...
+    """
+    Signal handler function to catch cancellation signals.
+    """
+    print("Received cancellation signal. Cleaning up...", signum)
+    # Add cleanup actions here, such as saving model state or logging metrics
+    # For example, you can log an intermediate metric to MLflow before exiting
+    if use_mlflow: 
+        mlflow.end_run("KILLED")
+    # Exit gracefully after cleanup
+    exit(1)
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: is this necessary? If you exit here, will other signal handlers also execute (there may be handlers deeper down the software stack)? I'm not sure how `exit` works under the hood ...

File: SynthSeg/training_with_tfrecords.py [hunk 6]
```diff
@@ -33,7 +47,11 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
     # Check epochs
     print(f"Setting seed to {opts.seed}")
     tf.keras.utils.set_random_seed(seed=opts.seed)
-
+    
+    #this is to monitor and end mlflow job with "Kinned" status in case job finishes to early - e.g. due to time limitations
+    signal.signal(signal.SIGINT, functools.partial(signal_handler, opts.mlflow))
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: I would move this into a `if opts.mlflow` block, see suggestion below.

File: SynthSeg/mlflow_utils.py
Comment (dcfidalgo) [inline SynthSeg/mlflow_utils.py:1]: Maybe we could be more specific with the naming, simply calling the file `mlflow_callback.py`.

File: SynthSeg/training_with_tfrecords.py [hunk 7]
```diff
@@ -50,6 +69,12 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
         strategy = tf.distribute.MirroredStrategy()
     else:
         raise NotImplementedError(f"The '{opts.strategy}' strategy is not implemented.")
+    
+    
+    if opts.mlflow:
+        from .mlflow_utils import set_mlflow_tracking
+        set_mlflow_tracking()
+        mlflow.start_run(log_system_metrics=True)
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]: Taking into account the other comments, i would suggest something like this
```suggestion
        import mlflow
        
        def handler(signum, frame):
            print("Received cancellation signal. Cleaning up...", signum)
            # Add cleanup actions here, such as saving model state or logging metrics
            # For example, you can log an intermediate metric to MLflow before exiting
            mlflow.end_run("KILLED")
                
        mlflow.tensorflow.autolog(disable=True)        
        mlflow.start_run(log_system_metrics=True)
        signal.signal(signal.SIGINT, handler)
        signal.signal(signal.SIGTERM, handler)  
```

**Review by dcfidalgo — CHANGES_REQUESTED** [review]
:100: Nice! I left a few inline comments, I think the main one is to take advantage of the integrated MLFlow env variables to save a few lines of code.

File: scripts/slurm/mlflow_example/training.slurm [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+nvidia-smi --query-gpu=timestamp,utilization.gpu,utilization.memory --format=csv -l 2 > nvidia_smi_monitoring.csv &
+NVIDIASMI_PID=$!
+
+apptainer_opts=(
+  --nv
+  -B $code_dir,$data_dir,$output_dir
+  --env PYTHONPATH=$code_dir
+  # in the container this is set to true by default
+  --env TF_FORCE_GPU_ALLOW_GROWTH=true
+  --env MLFLOW_CONFIG="$HOME/Projects/SynthSeg/scripts/slurm/mlflow_example/config.ini"
```
Comment (dcfidalgo) [inline scripts/slurm/mlflow_example/training.slurm]: Instead of defining the env variable, simply source the `mlflow_env.sh` script before executing the training script.

File: scripts/slurm/mlflow_example/config.ini
Comment (nhorlava) [reply scripts/slurm/mlflow_example/config.ini:1]: good idea, done!

File: SynthSeg/mlflow_utils.py (hunk 1 above)
Comment (nhorlava) [reply SynthSeg/mlflow_utils.py]: done:)

File: SynthSeg/training_options.py (hunk 2 above)
Comment (nhorlava) [reply SynthSeg/training_options.py]: yeah makes sense, done:)

File: SynthSeg/training_with_tfrecords.py (hunk 5 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: hmm not sure, I removed the exit line and works fine, so I guess we should leave it without it as you suggested below

File: SynthSeg/mlflow_utils.py
Comment (nhorlava) [reply SynthSeg/mlflow_utils.py:1]: since I now removed the setup_tracking function makes sense to rename it, done:)

File: SynthSeg/training_with_tfrecords.py (hunk 7 above)
Comment (nhorlava) [reply SynthSeg/training_with_tfrecords.py]: Nice suggestion! Implemented and tested it, works fine:)

File: scripts/slurm/mlflow_example/training.slurm (hunk 8 above)
Comment (nhorlava) [reply scripts/slurm/mlflow_example/training.slurm]: done -  I only put mlflow-related environmental variables there since I think it might be better to separate them from the general ones - what do you think?

**Comment by nhorlava** [comment]
@dcfidalgo addressed your comments in the last commit :)

File: SynthSeg/mlflow_callback.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        
+        
+        metrics = {"val_" + k: v for k, v in self.transform_logs(logs).items()}
+        mlflow.log_metrics(metrics,  step=current_iteration, synchronous=False)
+    
+        
+    def transform_logs(self, logs):
+        return {f"{k}-{self.metric_type}" if ("loss" in k) else k: v for k, v in logs.items()} if self.metric_type is not None else logs
+       
+
```
Comment (dcfidalgo) [inline SynthSeg/mlflow_callback.py]:
```suggestion
```

File: SynthSeg/training_with_tfrecords.py [hunk 10]
```diff
@@ -15,15 +15,14 @@
 
 from .metrics_model import WeightedL2Loss, DiceLoss, IdentityLoss, MeanIoU
 from .training_options import TrainingOptions
-from .brain_generator import read_tfrecords
-
+from .brain_generator import read_tfrecords
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion
from .brain_generator import read_tfrecords
```

File: SynthSeg/training_with_tfrecords.py [hunk 11]
```diff
@@ -15,15 +15,14 @@
...
+from .brain_generator import read_tfrecords 
+import functools
 
 class NullStrategy:
     @staticmethod
     def scope():
         return nullcontext()
-
-
+
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py:25]:
```suggestion
         
         
```

File: SynthSeg/training_with_tfrecords.py [hunk 12]
```diff
@@ -33,7 +32,7 @@ def training(opts: TrainingOptions) -> tf.keras.callbacks.History:
     # Check epochs
     print(f"Setting seed to {opts.seed}")
     tf.keras.utils.set_random_seed(seed=opts.seed)
-
+
```
Comment (dcfidalgo) [inline SynthSeg/training_with_tfrecords.py]:
```suggestion

```

**Review by dcfidalgo — APPROVED** [review]
perfect :rocket: Thanks a lot! I will merge this.
